rfcs/20250910-graph-api-dropout/README.md

```diff
@@ -0,0 +1,183 @@
+ONNX's Dropout operation [[#8]][8] takes `data`, `ratio` (probability) and `training_mode`
+as inputs, and `seed` as an attribute. Outputs include the dropped tensor and
+the dropout mask. ONNX does not specify the algorithm used to generate random numbers.
+
+**cuDNN**
+
+cuDNN Graph library does not provide a dedicated Dropout operation. Instead, it
+introduces an RNG operation, which can be used to implement Dropout with two
+pointwise Multiply operations.
+The RNG operation takes `seed`, `offset` as inputs, supports various
+distributions (Bernoulli, Uniform, Normal) by setting attributes, and outputs
+the generated tensor. The RNG is based on the Philox algorithm.
+
+**oneDNN**
+
+In oneDNN primitive, Dropout is implemented as a binary post-op applied to
+output tensor before other post-ops. It uses the Philox algorithm and takes
+`seed` (s32) and `probability` (f32) as inputs, and outputs a `mask` (u8).
+
+## Proposal
+
+Based on the investigation above, we propose to define the Dropout operation in
+the oneDNN Graph API as follows:
+
+**Operation Kind:**
+
+- `Dropout` (C++), `dnnl_graph_op_dropout` (C)
+
+**Inputs:**
+
+- `src`: input tensor (`f32` / `bf16` / `f16`)
+- `seed`: random seed (`uint64_t`, single value)
+- `offset`: random offset (`uint64_t`, single value)
+- `p`: dropout probability (`f32`, single value)
+
+**Outputs:**
+
+- `dst`: output tensor (`f32` / `bf16` / `f16`)
+- `mask`: mask tensor (`u8`, optional)
+
+**Notes:**
+
+- Dropout will use the Philox algorithm for random number generation.
+- Data types for `src`, `p`, `dst` and `mask` are consistent with the
+  current oneDNN primitive design.
+- The `mask` output is optional, allowing flexibility for scenarios where
+  the mask is not needed.
+- The `offset` input is introduced to ensure reproducibility and correct
+  partitioning across devices and threads, following best practices from PyTorch
+  and cuDNN.
+- Data types for `seed` and `offset` are aligned to PyTorch to ensure
+  compatibility and support large models.
+- A dedicated DropoutBackward operation is not defined, as the backward
+  computation shares the same mathematical formula as the forward pass.
+
+**Example Usage:**
+
+```cpp
+
+using namespace dnnl::graph;
+
+graph g = graph(engine::kind::gpu);
+
+logical_tensor src = logical_tensor(ID_SRC, data_type::f16, {256, 1000});
+logical_tensor seed = logical_tensor(ID_SEED, data_type::u64, {1});
+logical_tensor offset = logical_tensor(ID_SEED, data_type::u64, {1});
+logical_tensor p = logical_tensor(ID_SEED, data_type::f32, {1});
+logical_tensor dst = logical_tensor(ID_DST, data_type::f16, {256, 1000});
+logical_tensor mask = logical_tensor(ID_MASK, data_type::u8, {256, 1000});
+
+op dropout = op(ID_DROPOUT, op::kind::Dropout, "dropout");
+dropout.add_input(src);
+dropout.add_input(seed);
+dropout.add_input(offset);
+dropout.add_input(p);
+dropout.add_output(dst);
+dropout.add_output(mask);
+
+g.add_op(dropout);
+g.finalize();
+
+```
+
+**Open Question:**
```

Comment:
Random place: please consider and document how it will be tested, especially about the random-ness part.

Reply:
Thanks for the suggestion, will do.

Reply:
If we stick with Philox, the generation is actually deterministic: for a given output element at index idx, the generated rng value is Philox(seed, offset + idx).

The thing to be careful with is that the flattened `idx` should be consistent between reference and optimized impl (physical offset to base ptr should do).